### ISSUE TITLE
How should a11y string keys be named?

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/rosetta#201]
From a slack discussion, @zepumph asked what the convention should be for naming a11y strings. Should key names match the sentence/phrase of their value, or do we name them for their use? 

@jessegreenberg said that by name makes most sense because it will align with non-a11y string keys.

@mbarlow12 said that he has matched the value for shorter strings, but used thematic naming when the a11y strings get very long.

@pixelzoom wondered if having very long keys could present UI layout issues for Rosetta.

As an example, @zepumph brought up
```js
    grabbedLighlyOnBook: {
      value: 'Grabbed. Lightly on Physics book.'
    },
```
Should it be something else?
`grabbedLightlyOnPhysicsBook` (very long but matches string exactly)
`grabbedNotTouching` (thematic name to describe when this string is used)
Something else?

**Comment by jessegreenberg** [comment]
My understanding is that well named keys are mostly for translators to understand the context. So "perfect" keys (described in the code review checklist) are mostly for their benefit. In part of making a11y strings translatable, we will likely add context information about the string in the rosetta UI. If that happens, we don't need to provide context information in the string key.

But I think that the example @zepumph provided is good as is. I think it is OK to to remove words from the key to keep it shorter, but generally match the key to the string value. This matches non-a11y strings in the project.

**Comment by samreid** [comment]
Matching the English string seems like a straightforward and foolproof convention--I don't think we should second guess it or try to outsmart it.

> I think it is OK to to remove words from the key to keep it shorter

Generally that can cause problems.  If we remove words to end up with a string like `"grabbedLightlyOnBook"` then later there may be a Biology book and Chemistry book, so there would be asymmetry in the keys.  Simplest to just match the English words even if it is verbose.

**Comment by pixelzoom** [comment]
@samreid said:
> Matching the English string seems like a straightforward and foolproof convention

What about strings like this one in BASEA11yStrings:

```js
    simOpening: {
      value: 'The Play Area is a small room. The Control Panel has buttons, a checkbox, and radio buttons to change conditions in the room.'
    },
```

If we match the English string, that becomes:

```js
    the PlayAreaIsASmallRoomTheControlPanelHasButtonsACheckboxAndRadioButtonsToChangeConditions InTheRoom': {
      value: 'The Play Area is a small room. The Control Panel has buttons, a checkbox, and radio buttons to change conditions in the room.'
    },
```

Besides being a royal pain to create the key (which I can vouch for) and keep it in sync with the value, I can imagine that a key with that length might cause UI layout problems in Rosetta. 

So I think that a11y puts us in new territory, and we need to rethink our convention for naming keys.

**Comment by samreid** [comment]
That does seem like a special case, I'm fine if that needs a different convention, but we will need to know where to draw the line (in terms of string length).  Is that length of string frequent?  What if we had a javascript utility that took an English string and made a corresponding key, to do the busywork for us?  What if we had automatic validation of keys, to help us keep in line?

**Comment by samreid** [comment]
Testing in a context I borrowed from Molecule Shapes:
```js
if ( isFinite( centerDevicePoint.x ) && isFinite( centerDevicePoint.y ) ) {
  // screen coordinates
  var screenCenterPoint = new Vector2( ( centerDevicePoint.x + 1 ) * this.screenView.screenWidth / 2,
    ( -centerDevicePoint.y + 1 ) * this.screenView.screenHeight / 2 );
  var screenMidPoint = new Vector2( ( midDevicePoint.x + 1 ) * this.screenView.screenWidth / 2,
    ( -midDevicePoint.y + 1 ) * this.screenView.screenHeight / 2 );

  var labelString = Util.toFixed( angle, 1 ) + '°';
  while ( labelString.length < 5 ) {
    // handle single-digit labels by padding them
    labelString = '0' + labelString;
  }

  this.label.setLabel( 
    thePlayAreaIsASmallRoomTheControlPanelHasButtonsACheckboxAndRadioButtonsToChangeConditionsInTheRoom, 
    this.viewOpacity, screenCenterPoint, screenMidPoint, layoutScale );
}
```

At least it is very clear!  I don't feel strongly one way or the other.

**Comment by pixelzoom** [comment]
In the above example, the string's var name would actually be a tad longer, since a "String" suffix is required on all vars loaded via the `string!` plugin.

**Comment by pixelzoom** [comment]
Proposal conventions for string key naming:

(1) Key should typically be the value converted to camel case, e.g.:

```js
"helloWorld": { 
  value: "Hello World!" 
}
```

(2) If a key would be exceptionally long by following rule 1, use a key name that is an abbreviated form of the string value, or that captures the purpose/essence of the value. Examples:

```js
//key is abbreviated
"iWentToTheStore": {
  value: "I went to the store to get milk, eggs, butter, and sugar."
},

// key is based on purpose 
"describeTheRoom": { 
  value: "The Play Area is a small room. The Control Panel has buttons, a checkbox, and radio buttons to change conditions in the room." 
}
```

(3) If key names would collide, use your judgment to disambiguate. E.g.:

```js
"simplifyTitle": { 
   value: "Simplify!" 
},
"simplifyCheckbox": { 
   value: "simplify" 
}
```

**Comment by zepumph** [comment]
All of that sounds really nice. Potentially we could add a suffix for `2.A` such that it was know that the literal words of the key were just part of the actual value. 
`iWentToTheStoreAbbreviation`
`iWentToTheStorePart`
`iWentToTheStorePiece`
`iWentToTheStoreIntro`
iWentToTheStoreAbbrev`

I'm not totally sold; just a thought.

**Comment by pixelzoom** [comment]
12/13/18 dev meeting:

* proposal in https://github.com/phetsims/rosetta/issues/201#issuecomment-439478002 accepted as is
* I will copy the proposal to the code review checklist
